docs/design/rosbag2_playback_time.md

```diff
@@ -0,0 +1,200 @@
+  * If we are not able to contribute this feature into `rclcpp` in time for the Galactic API freeze - it may need to temporarily go into a rosbag2-based subclass for the Galactic release.
+
+```
+/**
+ * Used to control the timing of bag playback.
+ * This clock should be used to query times and sleep between message playing,
+ * so that the complexity involved around time control and time sources
+ * is encapsulated in this one place.
+ * Internally, it may own an rclcpp::Clock, but does not override the class in order to implement
+ * a slightly different API.
+ */
+class PlayerClock
+{
+public:
+  /**
+   * Constructor.
+   *
+   * \param use_sim_time
+   *   If false, starts paused.
+   *     Use member functions to configure before calling set_paused(false)
+   *   If true, subscribe to /clock to provide time
+   *     All time control (rate, pause, jump) is disabled, and will not publish to /clock
+   */
+  explicit PlayerClock(bool use_sim_time);
+
+  virtual ~PlayerClock();
+
+  // rclcpp::Clock interface
+  /**
+   * Provides the current time according to the clock's internal model.
+   *  if use_sim_time: provides current ROS Time (with optional extrapolation - see "Clock Rate and Time Extrapolation" section)
+   *  if !use_sim_time: calculates current "Player Time" based on starting time, playback rate, pause state.
+   *    this means that /clock time will match with the recorded messages time, as if we are fully reliving the recorded session
+   */
+  rclcpp::Time now();
+
+  /**
+   * Sleeps (non-busy wait) the current thread until the provided time is reached - according to this Clock
+   *
+   * If time is paused, the requested time may never be reached
+   * `real_time_timeout` uses the internal steady clock, if the timeout elapses, return false
+   * If jump() is called, return false, allowing the caller to handle the new time
+   * Return true if the time has been reached
+   */
+  bool sleep_until(rclcpp::Time until, rclcpp::Duration real_time_timeout);
+
+  // Time Control interface
+
+  /**
+   * Pauses/resumes time. Defaults to true (paused)
+   * While paused, `now()` will repeatedly return the same time, until resumed
+   *
+   * Note: this could have been implemented as `set_rate(0)`, but this interface allows this clock to maintain the
+   * clock's rate internally, so that the caller does not have to save it in order to resume.
+   */
+  void set_paused(bool paused);
+  bool get_paused() const;
+
+  /**
+    * Changes the rate of playback. Defaults to 1.0
+    * \param rate: must be nonzero positive. To pause playback, use \sa set_paused instead
+    * \throws std::invalid_argument if rate is <= 0.0
+    */
+  void set_rate(float rate);
+  float get_rate() const;
+
+  /**
+   * Set the rate that the clock will be published. Defaults to 0.
+   * \param frequency If this is set to <= 0, then the clock will not be published
+   */
+  void set_clock_publish_frequency(float frequency);
```

Comment:
Is this frequency in Hertz?  Is that defined as a standard somewhere in ros?

Reply:
Yes. I will clarify this, to make sure that we're definitely not talking about kilohertz or millihertz :)

Reply:
Does this start publishing immediately when frequency is set to >0?   Does the client need to remember to call jump to set the initial time before calling this method?

Reply:
if it's hertz, I'd recommend using a `size_t` data type or something non decimal.

Reply:
It's hertz, but it could be fractional Hz, 2.5 is a valid frequency

Reply:
fair enough.